### DISCUSSION TITLE
dbt and cyclical dependencies in Dagster

### DISCUSSION THREAD

**Discussion by bijzz** [dagster-io/dagster#13683] [Q&A] [answered]
Hi,

i have a workflow in which my `model_input` asset creates an `model_result` asset. In the real world scenario there are some assets before, in-between and afterwards but those have been left out for simplicity.

![image](https://user-images.githubusercontent.com/1084828/232555371-499e200f-976d-4128-87e1-c035e989f632.png)

Lets say they are daily partitioned. Now for another run today with `2023-04-17` our upstream asset `model_input` has to use the previous day data `2023-04-16` of the downstream asset `model_result`. 

In DBT itself this is not directly possible. When using `{{ ref('model_result') }}` within the `model_input` asset DBT will tell us `RuntimeError: Found a cycle: model.ml_use_case.model_result --> model.ml_use_case.model_input`. One workaround is to trick DBT in using the same downstream table twice - once as a reference and once as a source entity - and then perform the previous day calculation within SQL.

So using `{{ source('ml_use_case', 'model_result_input') }}` and define `sources.yml` as 

```
version: 2

sources:
  - name: ml_use_case
    schema: ml_use_case
    database: gcp-project-id
    tables:
      - name: model_result_input
        identifier: model_result
```
In Dagster we would end up with the following image in which `model_result_input` acts as a SourceAsset.

![image](https://user-images.githubusercontent.com/1084828/232558157-07b1e782-3c1b-4490-8da4-da54f3e80754.png)

Likely, one could somehow work with that but it looks and feels like a workaround. From the visualization it is not obvious to know that `model_result_input` is the same asset as `model_result`.

Now, to my understanding Dagster does allow the ability that upstream assets are able to utilize previous partitions of downstream assets. https://dagster.io/blog/partitioned-data-pipelines talks about "Depending on earlier or later partitions". I also found https://github.com/dagster-io/dagster/discussions/11829 talking about "How to write an asset that depends on an earlier partition of itself". 

My question is (A) is it possible to have an upstream asset depend on an earlier partition of an downstream asset? and (B) can this be accomplished when importing a DBT project? The https://github.com/dagster-io/dagster/discussions/11829 uses a partition mapping on an asset. But i do not think its possible to change this setting for an DBT imported asset of `load_assets_from_dbt_project`?

Any advice on the best practice / approach for my endeavor is much appreciated.

**Comment by OwenKephart** [comment, marked as answer]
Hi @bijzz ! Good question. Dagster does support some level of cyclical dependencies, but these are only for assets which depend on earlier versions of themselves (i.e. a table that uses its past three days of data to compute the next day). In the example above, the model_result asset *transitively* depends on itself, which is not currently supported in Dagster, and is not currently scoped / planned.

So, for your case, I believe the best solution is still along the lines of what you've illustrated -- essentially using two Dagster assets to represent the same underlying asset.
